### PR TITLE
add simple shard computer

### PR DESCRIPTION
**Pull request by Penglq**
**What this PR does**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/runtime/rule/shard.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+)
+
+var shardMap = map[ShardType]fn{
+	ModShard:       NewModShard,
+	HashMd5Shard:   NewHashMd5Shard,
+	HashCrc32Shard: NewHashCrc32Shard,
+}
+
+func ShardFactory(shardType ShardType, shardNum int) (shardStrategy rule.ShardComputer, err error) {
+	if f, ok := shardMap[shardType]; ok {
```
Comment (jjeffcaii) [inline pkg/runtime/rule/shard.go:52]: add validation: shardNum > 0

File: pkg/runtime/rule/shard.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	md5Hash := md5.New()
+	md5Hash.Write([]byte(s))
+	md5Str := hex.EncodeToString(md5Hash.Sum(nil))
+
+	h := 0
+	byteHash := []byte(md5Str)
+	for i := 0; i < 32; i++ {
+		h <<= 8
+		h |= int(byteHash[i]) & 0xFF
+	}
```
Comment (jjeffcaii) [inline pkg/runtime/rule/shard.go:110]: The length of md5 bytes is 16, see `md5.Size`, use it directly, I think it's not necessary to convert it as hex string.

File: pkg/util/maths/abs.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+
+package maths
+
+// Abs abs
+func Abs(n int64) int64 {
```
Comment (AlexStocks) [inline pkg/util/maths/abs.go:22]: 像这种函数，挪到 gost 库 math 里面去。。

File: pkg/runtime/rule/shard.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+func ShardFactory(shardType ShardType, shardNum int) (shardStrategy rule.ShardComputer, err error) {
+	if f, ok := shardMap[shardType]; ok {
+		return f(shardNum), nil
+	}
+	return nil, errors.New("do not have this shardType")
+}
+
+type ShardType string
+type fn func(shardNum int) rule.ShardComputer
```
Comment (AlexStocks) [inline pkg/runtime/rule/shard.go:59]: 把 fn 改成 ShardComputerFunc 吧

**Review by jjeffcaii — APPROVED** [review]
LGTM

**Review by AlexStocks — APPROVED** [review]

**Review by AlexStocks — APPROVED** [review]

**Review by liu-song — APPROVED** [review]

**Pull request merged**